File: leetcode/Java/49_Group_Anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // 애너그램 결과 보관
        Map<String, List<String>> results = new HashMap<>();

        // 입력값 문자열 배열 순회
        for (String s : strs){
            // 문자열을 문자 배열로 변환
            char[] chars = s.toCharArray();
            // 문자 배열 정렬
            Arrays.sort(chars);
            // 문자 배열을 키로 하기 위해 다시 문자열로
            String key = String.valueOf(chars);

            //기존x 키라면 빈 리스트 삽입
            if(!results.containsKey(key))
                results.put(key, new ArrayList<>());
            //키에 해당하는 리스트에 추가
            results.get(key).add(s)
;        }
        return new ArrayList<>(results.values());
    }
}

File: leetcode/Kotlin/49_Group_Anagrams.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        // 애너그램 결과 보관
        val results:MutableMap<String,MutableList<String>> = mutableMapOf();

        // 입력값 문자열 배열 순회
        for (s in strs) {
            // 문자열 정렬
            val key = s.toCharArray().sorted().joinToString("")
            // 기존에 없던 키라면 빈 리스트 삽입
            results.getOrPut(key){ mutableListOf() }
            // 키에 해당하는 리스트에 추가
            results[key]!!.add(s)
        }
        return ArrayList<List<String>>(results.values)
    }
}
